// express-insights/htmlRenderer.js
const sanitizeHtml = require('sanitize-html');

function renderHTML(data, message, style, autoRefreshInterval) {
  const isDark = style === 'dark';
  const theme = isDark
    ? {
        bg: '#0d1117',
        cardBg: '#161b22',
        text: '#f0f6fc',
        border: '#30363d',
        success: '#2ea043',
        error: '#da3633',
        warning: '#fb8500',
        accent: '#58a6ff',
      }
    : {
        bg: '#f6f8fa',
        cardBg: '#ffffff',
        text: '#24292f',
        border: '#d0d7de',
        success: '#1a7f37',
        error: '#cf222e',
        warning: '#bf8700',
        accent: '#0969da',
      };

  const statusColor =
    data.status === 'healthy' ? theme.success : data.status === 'degraded' ? theme.warning : theme.error;

  return `<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Health Check - ${sanitizeHtml(data.status)}</title>
  <style>
    * { margin: 0; padding: 0; box-sizing: border-box; }
    body { font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif; background: ${theme.bg}; color: ${theme.text}; line-height: 1.6; padding: 20px; }
    .container { max-width: 1200px; margin: 0 auto; }
    .header { text-align: center; margin-bottom: 30px; padding: 20px; background: ${theme.cardBg}; border-radius: 10px; border: 1px solid ${theme.border}; }
    .status-badge { display: inline-block; padding: 5px 15px; border-radius: 20px; font-weight: bold; text-transform: uppercase; color: white; background: ${statusColor}; margin-left: 10px; }
    .grid { display: grid; grid-template-columns: repeat(auto-fit, minmax(300px, 1fr)); gap: 20px; margin-bottom: 30px; }
    .card { background: ${theme.cardBg}; border: 1px solid ${theme.border}; border-radius: 10px; padding: 20px; box-shadow: 0 2px 4px rgba(0,0,0,0.1); }
    .card h3 { color: ${theme.accent}; margin-bottom: 15px; font-size: 1.2em; }
    .metric { display: flex; justify-content: space-between; margin-bottom: 8px; padding: 5px 0; border-bottom: 1px solid ${theme.border}; }
    .metric:last-child { border-bottom: none; }
    .metric-label { font-weight: 500; }
    .metric-value { font-family: 'Monaco', 'Menlo', monospace; font-size: 0.9em; }
    .json-container { background: ${theme.cardBg}; border: 1px solid ${theme.border}; border-radius: 10px; padding: 20px; overflow-x: auto; }
    pre { font-family: 'Monaco', 'Menlo', monospace; font-size: 0.85em; line-height: 1.4; white-space: pre-wrap; }
    .refresh-btn { position: fixed; top: 20px; right: 20px; background: ${theme.accent}; color: white; border: none; padding: 10px 20px; border-radius: 5px; cursor: pointer; font-weight: bold; }
    .refresh-btn:hover { opacity: 0.8; }
    .timestamp { font-size: 0.9em; color: ${theme.text}88; margin-top: 10px; }
  </style>
</head>
<body>
  <button class="refresh-btn" onclick="window.location.reload()">🔄 Refresh</button>
  <div class="container">
    <div class="header">
      <h1>${sanitizeHtml(message)}</h1>
      <span class="status-badge">${sanitizeHtml(data.status)}</span>
      <div class="timestamp">Last updated: ${sanitizeHtml(data.timestamp.toISOString())}</div>
    </div>
    <div class="grid">
      <div class="card">
        <h3>📊 System Overview</h3>
        <div class="metric"><span class="metric-label">Uptime</span><span class="metric-value">${sanitizeHtml(data.uptime)}</span></div>
        <div class="metric"><span class="metric-label">Environment</span><span class="metric-value">${sanitizeHtml(data.environment)}</span></div>
        <div class="metric"><span class="metric-label">Version</span><span class="metric-value">${sanitizeHtml(data.version)}</span></div>
        <div class="metric"><span class="metric-label">Platform</span><span class="metric-value">${sanitizeHtml(data.system.platform)}</span></div>
      </div>
      <div class="card">
        <h3>💾 Memory Usage</h3>
        <div class="metric"><span class="metric-label">Heap Used</span><span class="metric-value">${sanitizeHtml(data.memory.heapUsed)} (${sanitizeHtml(data.memory.heapUsedPercent)})</span></div>
        <div class="metric"><span class="metric-label">Heap Total</span><span class="metric-value">${sanitizeHtml(data.memory.heapTotal)}</span></div>
        <div class="metric"><span class="metric-label">RSS</span><span class="metric-value">${sanitizeHtml(data.memory.rss)}</span></div>
        <div class="metric"><span class="metric-label">External</span><span class="metric-value">${sanitizeHtml(data.memory.external)}</span></div>
      </div>
      <div class="card">
        <h3>🖥️ System Resources</h3>
        <div class="metric"><span class="metric-label">CPU Cores</span><span class="metric-value">${data.system.cpus.count}</span></div>
        <div class="metric"><span class="metric-label">Load Average</span><span class="metric-value">${sanitizeHtml(data.system.loadavg['1min'])}</span></div>
        <div class="metric"><span class="metric-label">System Memory</span><span class="metric-value">${sanitizeHtml(data.system.memory.used)} / ${sanitizeHtml(data.system.memory.total)}</span></div>
        <div class="metric"><span class="metric-label">Memory Usage</span><span class="metric-value">${sanitizeHtml(data.system.memory.usedPercent)}</span></div>
      </div>
      ${data.services
        ? `
      <div class="card">
        <h3>🔧 Service Status</h3>
        ${Object.entries(data.services)
          .map(
            ([name, status]) => `
          <div class="metric">
            <span class="metric-label">${sanitizeHtml(name)}</span>
            <span class="metric-value" style="color: ${status === 'healthy' ? theme.success : theme.error}">
              ${status === 'healthy' ? '✅' : '❌'} ${sanitizeHtml(status)}
            </span>
          </div>
        `
          )
          .join('')}
      </div>
      `
        : ''}
    </div>
    <div class="json-container">
      <h3>📋 Raw Data</h3>
      <pre>${sanitizeHtml(JSON.stringify(data, null, 2))}</pre>
    </div>
  </div>
  <script>
    ${autoRefreshInterval ? `setTimeout(() => window.location.reload(), ${autoRefreshInterval});` : ''}
  </script>
</body>
</html>`;
}

module.exports = { renderHTML };
